src/protocol.cpp

```diff
@@ -90,6 +92,10 @@ const static std::vector<std::string> g_all_net_message_types{
     NetMsgType::SENDTXRCNCL,
 };
 
+static_assert(NUM_NET_MESSAGE_TYPES == sizeof(allNetMessageTypes) / sizeof(allNetMessageTypes[0]), "Please update NUM_NET_MESSAGE_TYPES");
+
+const static std::vector<std::string> allNetMessageTypesVec(std::begin(allNetMessageTypes), std::end(allNetMessageTypes));
```

Comment:
All these plus the `NUM_NET_MESSAGE_TYPES` constant can be avoided if the list of all message types is turned into a `std::array`. Consider this:

<details>
<summary>[patch] make the list of known message types a compile time constant</summary>

```diff
diff --git a/src/net.cpp b/src/net.cpp
index a2f80cbcf7..133abae117 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -3670,13 +3670,13 @@ CNode::CNode(NodeId idIn,
       nLocalHostNonce{nLocalHostNonceIn},
       m_recv_flood_size{node_opts.recv_flood_size},
       m_i2p_sam_session{std::move(node_opts.i2p_sam_session)}
 {
     if (inbound_onion) assert(conn_type_in == ConnectionType::INBOUND);
 
-    for (const std::string &msg : getAllNetMessageTypes())
+    for (const auto& msg : g_all_net_message_types)
         mapRecvBytesPerMsgType[msg] = 0;
     mapRecvBytesPerMsgType[NET_MESSAGE_TYPE_OTHER] = 0;
 
     if (fLogIPs) {
         LogPrint(BCLog::NET, "Added connection to %s peer=%d\n", m_addr_name, id);
     } else {
diff --git a/src/protocol.cpp b/src/protocol.cpp
index 27a0a2ffc1..0b22cc47aa 100644
--- a/src/protocol.cpp
+++ b/src/protocol.cpp
@@ -46,53 +46,12 @@ const char* CFHEADERS = "cfheaders";
 const char* GETCFCHECKPT = "getcfcheckpt";
 const char* CFCHECKPT = "cfcheckpt";
 const char* WTXIDRELAY = "wtxidrelay";
 const char* SENDTXRCNCL = "sendtxrcncl";
 } // namespace NetMsgType
 
-/** All known message types. Keep this in the same order as the list of
- * messages above and in protocol.h.
- */
-const static std::vector<std::string> g_all_net_message_types{
-    NetMsgType::VERSION,
-    NetMsgType::VERACK,
-    NetMsgType::ADDR,
-    NetMsgType::ADDRV2,
-    NetMsgType::SENDADDRV2,
-    NetMsgType::INV,
-    NetMsgType::GETDATA,
-    NetMsgType::MERKLEBLOCK,
-    NetMsgType::GETBLOCKS,
-    NetMsgType::GETHEADERS,
-    NetMsgType::TX,
-    NetMsgType::HEADERS,
-    NetMsgType::BLOCK,
-    NetMsgType::GETADDR,
-    NetMsgType::MEMPOOL,
-    NetMsgType::PING,
-    NetMsgType::PONG,
-    NetMsgType::NOTFOUND,
-    NetMsgType::FILTERLOAD,
-    NetMsgType::FILTERADD,
-    NetMsgType::FILTERCLEAR,
-    NetMsgType::SENDHEADERS,
-    NetMsgType::FEEFILTER,
-    NetMsgType::SENDCMPCT,
-    NetMsgType::CMPCTBLOCK,
-    NetMsgType::GETBLOCKTXN,
-    NetMsgType::BLOCKTXN,
-    NetMsgType::GETCFILTERS,
-    NetMsgType::CFILTER,
-    NetMsgType::GETCFHEADERS,
-    NetMsgType::CFHEADERS,
-    NetMsgType::GETCFCHECKPT,
-    NetMsgType::CFCHECKPT,
-    NetMsgType::WTXIDRELAY,
-    NetMsgType::SENDTXRCNCL,
-};
-
 CMessageHeader::CMessageHeader(const MessageStartChars& pchMessageStartIn, const char* pszCommand, unsigned int nMessageSizeIn)
     : pchMessageStart{pchMessageStartIn}
 {
     // Copy the command name
     size_t i = 0;
     for (; i < COMMAND_SIZE && pszCommand[i] != 0; ++i) pchCommand[i] = pszCommand[i];
@@ -175,17 +134,12 @@ std::string CInv::ToString() const
         return strprintf("%s %s", GetCommand(), hash.ToString());
     } catch(const std::out_of_range &) {
         return strprintf("0x%08x %s", type, hash.ToString());
     }
 }
 
-const std::vector<std::string> &getAllNetMessageTypes()
-{
-    return g_all_net_message_types;
-}
-
 /**
  * Convert a service flag (NODE_*) to a human readable string.
  * It supports unknown service flags which will be returned as "UNKNOWN[...]".
  * @param[in] bit the service flag is calculated as (1 << bit)
  */
 static std::string serviceFlagToStr(size_t bit)
diff --git a/src/protocol.h b/src/protocol.h
index e405253632..2a6b85dea5 100644
--- a/src/protocol.h
+++ b/src/protocol.h
@@ -264,14 +264,50 @@ extern const char* WTXIDRELAY;
  * The salt is used to compute short txids needed for efficient
  * txreconciliation, as described by BIP 330.
  */
 extern const char* SENDTXRCNCL;
 }; // namespace NetMsgType
 
-/* Get a vector of all valid message types (see above) */
-const std::vector<std::string>& getAllNetMessageTypes();
+/** All known message types (see above). Keep this in the same order as the list of messages above. */
+static const std::array g_all_net_message_types{
+    NetMsgType::VERSION,
+    NetMsgType::VERACK,
+    NetMsgType::ADDR,
+    NetMsgType::ADDRV2,
+    NetMsgType::SENDADDRV2,
+    NetMsgType::INV,
+    NetMsgType::GETDATA,
+    NetMsgType::MERKLEBLOCK,
+    NetMsgType::GETBLOCKS,
+    NetMsgType::GETHEADERS,
+    NetMsgType::TX,
+    NetMsgType::HEADERS,
+    NetMsgType::BLOCK,
+    NetMsgType::GETADDR,
+    NetMsgType::MEMPOOL,
+    NetMsgType::PING,
+    NetMsgType::PONG,
+    NetMsgType::NOTFOUND,
+    NetMsgType::FILTERLOAD,
+    NetMsgType::FILTERADD,
+    NetMsgType::FILTERCLEAR,
+    NetMsgType::SENDHEADERS,
+    NetMsgType::FEEFILTER,
+    NetMsgType::SENDCMPCT,
+    NetMsgType::CMPCTBLOCK,
+    NetMsgType::GETBLOCKTXN,
+    NetMsgType::BLOCKTXN,
+    NetMsgType::GETCFILTERS,
+    NetMsgType::CFILTER,
+    NetMsgType::GETCFHEADERS,
+    NetMsgType::CFHEADERS,
+    NetMsgType::GETCFCHECKPT,
+    NetMsgType::CFCHECKPT,
+    NetMsgType::WTXIDRELAY,
+    NetMsgType::SENDTXRCNCL,
+};
 
 /** nServices flags */
 enum ServiceFlags : uint64_t {
     // NOTE: When adding here, be sure to update serviceFlagToStr too
     // Nothing
     NODE_NONE = 0,
diff --git a/src/test/fuzz/p2p_transport_serialization.cpp b/src/test/fuzz/p2p_transport_serialization.cpp
index 21d8dab536..853661b4d4 100644
--- a/src/test/fuzz/p2p_transport_serialization.cpp
+++ b/src/test/fuzz/p2p_transport_serialization.cpp
@@ -18,19 +18,18 @@
 #include <limits>
 #include <optional>
 #include <vector>
 
 namespace {
 
-std::vector<std::string> g_all_messages;
+auto g_all_messages = g_all_net_message_types;
 
 void initialize_p2p_transport_serialization()
 {
     ECC_Start();
     SelectParams(ChainType::REGTEST);
-    g_all_messages = getAllNetMessageTypes();
     std::sort(g_all_messages.begin(), g_all_messages.end());
 }
 
 } // namespace
 
 FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serialization)
@@ -147,13 +146,13 @@ void SimulationTest(Transport& initiator, Transport& responder, R& rng, FuzzedDa
                 if (c < ' ' || c > 0x7E) break;
                 ret += c;
             }
             return ret;
         } else {
             // Otherwise, use it as index into the list of known messages.
-            return g_all_messages[v % g_all_messages.size()];
+            return std::string{g_all_messages[v % g_all_messages.size()]};
         }
     };
 
     // Function to construct a CSerializedNetMsg to send.
     auto make_msg_fn = [&](bool first) {
         CSerializedNetMsg msg;
diff --git a/src/test/fuzz/process_message.cpp b/src/test/fuzz/process_message.cpp
index d38d1bb40e..24c711f421 100644
--- a/src/test/fuzz/process_message.cpp
+++ b/src/test/fuzz/process_message.cpp
@@ -42,13 +42,13 @@ std::string_view LIMIT_TO_MESSAGE_TYPE{};
 } // namespace
 
 void initialize_process_message()
 {
     if (const auto val{std::getenv("LIMIT_TO_MESSAGE_TYPE")}) {
         LIMIT_TO_MESSAGE_TYPE = val;
-        Assert(std::count(getAllNetMessageTypes().begin(), getAllNetMessageTypes().end(), LIMIT_TO_MESSAGE_TYPE)); // Unknown message type passed
+        Assert(std::count(g_all_net_message_types.begin(), g_all_net_message_types.end(), LIMIT_TO_MESSAGE_TYPE)); // Unknown message type passed
     }
 
     static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>(
             /*chain_type=*/ChainType::REGTEST,
             /*extra_args=*/{"-txreconciliation"});
     g_setup = testing_setup.get();
```
</details>

Reply:
Thanks for the idea Vasil.

I implemented it in this [tag](https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:netmsgstats-vd-array) to see what it would look like.

Overall, I prefer the design, but in writing the commit message for the change I began to feel unsure that it _actually_ results in `g_all_net_message_types` being a compile-time constant.  IIUC the array will reference a `const char*`. The pointer addresses can only be allocated at runtime and in turn `g_all_net_message_types` will be static at runtime but not at compile time?

I'm not sure I see an easy way to have this be a compile time constant, if my understanding above is correct.

Reply:
Thanks, I forgot to mention I had tried that. It results in something like this:

```log
$ bear -- make
Making all in src
make[1]: Entering directory '/home/will/src/bitcoin/src'
make[2]: Entering directory '/home/will/src/bitcoin/src'
make[3]: Entering directory '/home/will/src/bitcoin'
make[3]: Leaving directory '/home/will/src/bitcoin'
  CXX      bitcoind-bitcoind.o
In file included from ./netstats.h:10,
                 from ./net.h:22,
                 from ./interfaces/node.h:10,
                 from ./interfaces/init.h:10,
                 from bitcoind.cpp:19:
./protocol.h:276:65: error: the type ‘const std::array<std::__cxx11::basic_string<char>, 35>’ of ‘constexpr’ variable ‘g_all_net_message_types’ is not literal
  276 | static constexpr std::array<std::string, NUM_NET_MESSAGE_TYPES> g_all_net_message_types{
      |                                                                 ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ./uint256.h:13,
                 from ./consensus/params.h:9,
                 from ./kernel/chainparams.h:9,
                 from ./chainparams.h:9,
                 from bitcoind.cpp:10:
/usr/include/c++/12/array:99:12: note: ‘std::array<std::__cxx11::basic_string<char>, 35>’ is not literal because:
   99 |     struct array
      |            ^~~~~
/usr/include/c++/12/array:99:12: note:   ‘std::array<std::__cxx11::basic_string<char>, 35>’ has a non-trivial destructor
make[2]: *** [Makefile:16350: bitcoind-bitcoind.o] Error 1
make[2]: Leaving directory '/home/will/src/bitcoin/src'
make[1]: *** [Makefile:20245: all-recursive] Error 1
make[1]: Leaving directory '/home/will/src/bitcoin/src'
make: *** [Makefile:813: all-recursive] Error 1
```

Reply:
`const std::array a{x, y, z};` - at compile time the compiler knows the size of the array. `a.size()` can be used to define the size of other arrays, e.g. `std::array<int, a.size()> anoher;`. This is what matters. Whether the contents of the array is a compile time constant (the pointers), I am not sure.